### PR TITLE
Wip 301 pure op crdts

### PR DESCRIPTION
**Pull request by gabrielgiussi**
Some notes:

- The implmentation is based on [Pure Operation-Based Replicated Data Types](https://arxiv.org/abs/1710.04469)
- All CRDTs, except Counter, were replaced by CvRDTPureOp. Counter is commutative so its design is simpler.
- AWSet is the only implemented with a custom state (i.e. a Set (1)). The state of the rest of the CRDTs is just a `Seq[Operation]`
- The stable state of SimpleCRDT (aka a CRDT which state is a `Seq[Operation]`) is a `Seq[Operation]` (because the VectorTime is no longer needed). This save us a VectorTime per operation in memory, but it forces to [create a Versioned with VectorTime.Zero per operation each time a new operation is delivered](https://github.com/RBMHTechnology/eventuate/compare/master...gabrielgiussi:wip-301-pure-op-crdts?expand=1#diff-58adce1662577f90b4e07e182e9f37e3R183), so is open to discussion.
- I have to remove the implicit ops in each service and replaced by [a fixed ops](https://github.com/RBMHTechnology/eventuate/compare/master...gabrielgiussi:wip-301-pure-op-crdts?expand=1#diff-ade0c796f3057eb228f6d882398cdcaeR77). If this solution is not acceptable I will try again to make it work with implicit, maybe you can give me a hint on that.

(1) An improvement will be to be able to define the Set implementation.

### CODE REVIEW

**Comment by gabrielgiussi** [comment]
I've created the class `EventuateMultiNodeSpec` to facilitate the tests. If you are not ok with it I could rewrite the tests without it.

**Comment by krasserm** [comment]
@gabrielgiussi thanks a lot for the PR. Please bear with me that a review might take some time ... Since they actually replace the existing CRDT implementations we will have to make sure that they all pass [eventuate-chaos](https://github.com/RBMHTechnology/eventuate-chaos) tests. @volkerstampa can you setup the chaos tests for the new implementation? Alternatively, if you want to have them merged earlier so that users can start to experiment with, I propose the old implementation and the new pure op-based to co-exist for a while. WDYT?

**Comment by gabrielgiussi** [comment]
> we will have to make sure that they all pass eventuate-chaos tests

I could give it a try if you want.

> I propose the old implementation and the new pure op-based to co-exist for a while

I agree, I will reset the old classes as they were and move the new implementations to a package `pureop`.

**Comment by volkerstampa** [comment]
Currently the chaos tests are mainly testing eventuate itself and not so much the CRDTs. The only existing test-client uses the CRDT-counter. If we want to chaos-test each CRDT implementation corresponding test-clients would have to be written first. I can assist when it comes to implementing those but atm I cannot take care of the implementation by myself.

@gabrielgiussi It would be great if you could give it a try.

**Comment by krasserm** [comment]
@gabrielgiussi thanks for giving it a try! Regarding a separate implementation, what about adding a new module e.g. `eventuate-crdt-pure`?

**Comment by gabrielgiussi** [comment]
I'll add a new module, but I did some minor changes to `CRDTActor` and `CRDTManager` to implement stability, so the question is, you want me to left `[crdt]` module as it was and duplicate the `CRDTService`, `CRDTActor` and `CRDTManager` in `[crdt-pure]`? Or you are ok with some minor changes in `[crdt]` and make `[crdt-pure]` depend on it?

**Comment by gabrielgiussi** [comment]
I've run the counter scenario with eventuate 0.10 and I've got 
```
Processed 3956 requests in the meantime
All 3 nodes converged to the counter value: 2228
**Expected counter value: 2679; actual 2228**
```
It's ok?

**Comment by krasserm** [comment]
@gabrielgiussi Duplication is ok for the moment. It allows us to do further experiments without interfering with the existing implementation.

**Comment by krasserm** [comment]
@gabrielgiussi regarding the CRDT chaos runs: I remember that we had a bug in calculating the expected value but I thought it was fixed. That's a bug in eventuate-chaos itself. @volkerstampa have you encountered these issues as well?

**Comment by volkerstampa** [comment]
As long as the exit code of the test program is 0, the test is green, i.e. all nodes converged to the same value. The reported _expected_ counter value is actually only correct if all commands sent during the test can be processed successfully, however this is not guaranteed under chaos, so the _expectation_ is actually wrong and thus the report is sort of misleading.

**Comment by gabrielgiussi** [comment]
[This](https://github.com/gabrielgiussi/eventuate-chaos/commit/951e5aaa7b794d0414f72f8c00377632f05b6f5f) is the best I could do for now.

Also, I've already moved the implementation to a separate module, copying the code but changing the package to `com.rbmhtechnology.eventuate.crdt.pure` (added `pure`).

File: eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+class CRDTChaosSpecLeveldb extends WordSpec with Matchers with MultiLocationSpecLeveldb {
+  import ReplicationIntegrationSpec.replicationConnection
+  import CRDTChaosSpecLeveldb._
+
+  class TestEventLog(id: String) extends LeveldbEventLog(id, "log-test") {
+    override def write(events: Seq[DurableEvent], partition: Long, clock: EventLogClock): Unit =
+      if (events.map(_.payload).contains(ValueUpdated(AddOp(randomNr())))) throw IntegrationTestException else super.write(events, partition, clock)
+  }
+
+  // FIXME allow use write-batch-size > 1
```
Comment (volkerstampa) [inline eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala]: Is that still open? Do you still plan to fix this?

File: .travis.yml [hunk 2]
```diff
@@ -6,25 +6,19 @@ cache:
   directories:
     - $HOME/.ivy2/cache
     - $HOME/.sbt/boot/
-before_install:
```
Comment (volkerstampa) [inline .travis.yml]: Why is that removed?

**Comment by volkerstampa** [comment]
@gabrielgiussi The changes to eventuate-chaos in general look ok to me. Not sure if you want to open a PR for this as well?!? But probably only makes sense if this PR is merged ...

Thanks again for all your effort.

File: .travis.yml [hunk 3]
```diff
@@ -6,25 +6,19 @@ cache:
...
   - .travis/test-crdt.sh $TRAVIS_SCALA_VERSION
+  - .travis/test-crdt-pure.sh $TRAVIS_SCALA_VERSION
   - .travis/test-stream.sh $TRAVIS_SCALA_VERSION
   - .travis/test-spark.sh $TRAVIS_SCALA_VERSION
   - .travis/test-vertx.sh $TRAVIS_SCALA_VERSION
   - find $HOME/.sbt -name "*.lock" | xargs rm
   - find $HOME/.ivy2 -name "ivydata-*.properties" | xargs rm
 after_success:
-  - if [ "${TRAVIS_PULL_REQUEST}" = "false" ] && [ "${TRAVIS_BRANCH}" = "master" ]; then .travis/publish-all.sh $TRAVIS_SCALA_VERSION; fi
+- sbt ++$TRAVIS_SCALA_VERSION publish
```
[the remark this answers is not in the document]
Comment (gabrielgiussi) [reply .travis.yml]: I've modified the travis.yml to publish my own artifacts but I shouldn't have included in this PR.

File: eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/VectorTime.scala [hunk 4]
```diff
@@ -79,6 +79,9 @@ case class VectorTime(value: Map[String, Long] = Map.empty) {
   def <->(that: VectorTime): Boolean =
     conc(that)
 
+  def ||(that: VectorTime): Boolean =
```
[the remark this answers is not in the document]
Comment (gabrielgiussi) [reply eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/VectorTime.scala]: I've included `||` initially because is how is represented in literature but I didn't replace the reference to `<->`. If you agree with me that `||` is better I could update the only reference that exists to `<->`, if not I'll delete it.

File: build.sbt [hunk 5]
```diff
@@ -9,12 +9,12 @@ import ProjectDependencies._
 
 version in ThisBuild := "0.10-SNAPSHOT"
 
-organization in ThisBuild := "com.rbmhtechnology"
+organization in ThisBuild := "oss.gabrielgiussi"
```
[the remark this answers is not in the document]
Comment (gabrielgiussi) [reply build.sbt]: You guess wrong. Martin doesn't told you?  :smirk:
Same as above, I've updated the organization to publish my own artifacts to jfrog so the eventuate-chaos build could download them.

File: eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala (hunk 1 above)
Comment (gabrielgiussi) [reply eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala]: I've tried without success with 
```
class TestEventLog(id: String) extends LeveldbEventLog(id, "log-test") {
    override def write(events: Seq[DurableEvent], partition: Long, clock: EventLogClock): Unit = {
      val containsStop = events.map(_.payload).exists {
        case ValueUpdated(AddOp(s)) => s.toString.startsWith("stop")
        case _                      => false
      }
      if (!containsStop && events.map(_.payload).contains(ValueUpdated(AddOp(randomNr())))) throw IntegrationTestException else super.write(events, partition, clock)
    }
  }
```
I can't see what I'm missing, so I could let the `FIXME` as a reminder or delete it.

**Comment by gabrielgiussi** [comment]
If this PR is eventually merged I will open a PR for eventuate-chaos.

File: build.sbt (hunk 5 above)
[the remark this answers is not in the document]
Comment (krasserm) [reply build.sbt]: > You guess wrong. Martin doesn't told you?

Looks like I miss something. Can you please enlighten me?

[the remark this answers is not in the document]
Comment (gabrielgiussi) [reply build.sbt]: Sorry Martin, I'm not very good with jokes.

These changes shouldn't have been included in the PR, I just changed the organization to publish my own artifacts to jfrog so they will be available to the eventuate-chaos travis build.

File: eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/VectorTime.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (volkerstampa) [reply eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/VectorTime.scala]: I propose to stick with the _old_ symbol as `||` looks too much like a boolean or

File: eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala (hunk 1 above)
Comment (volkerstampa) [reply eventuate-crdt-pure/src/it/scala/com/rbmhtechnology/eventuate/crdt/CRDTChaosSpecLeveldb.scala]: For me both is fine.

**Comment by volkerstampa** [comment]
@gabrielgiussi Sorry for the really looong delay. are you still interested in merging this PR? It seems like neither Martin nor me will have time for a thorough review. As the new code is in a separate module the proposal is that the PR can be merged, once the small issues outside of the new module are fixed and you could consider yourself the owner of the new module. WDYT?

**Comment by gabrielgiussi** [comment]
@volkerstampa thanks for taking time to review this PR. If you think it helps to the project I will be happy to merge it and be the owner of the module.

**Comment by volkerstampa** [comment]
@gabrielgiussi Thanks for the contribution and special thanks for your patience. Please squash all commits and I will merge the PR.
